File: src/fos/continuity.py
"""FOS Continuity Engine."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any
from uuid import uuid4

from src.fos.types import ContinuityThreadId, Id


@dataclass
class ContinuityEvent:
    thread: ContinuityThreadId
    event_id: Id
    kind: str
    payload: dict[str, Any]

    def to_dict(self) -> dict[str, Any]:
        return {
            "thread": self.thread,
            "event_id": self.event_id,
            "kind": self.kind,
            "payload": dict(self.payload),
        }


class ContinuityEngine:
    @staticmethod
    def emit(kind: str, thread: ContinuityThreadId, payload: dict[str, Any]) -> ContinuityEvent:
        return ContinuityEvent(
            thread=thread,
            event_id=f"evt-{uuid4()}",
            kind=kind,
            payload=payload,
        )

File: fos-kernel/src/continuity.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::types::{ContinuityThreadId, Id};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ContinuityEvent {
    pub thread: ContinuityThreadId,
    pub event_id: Id,
    pub kind: String,
    pub payload: serde_json::Value,
}

pub struct ContinuityEngine;

impl ContinuityEngine {
    pub fn emit(
        kind: &str,
        thread: &ContinuityThreadId,
        payload: serde_json::Value,
    ) -> ContinuityEvent {
        ContinuityEvent {
            thread: thread.clone(),
            event_id: format!("evt-{}", Uuid::new_v4()),
            kind: kind.to_string(),
            payload,
        }
    }
}
